### PR TITLE
connlib: add client dns interception support

### PR DESCRIPTION
**Pull request by conectado**

### CODE REVIEW

File: rust/connlib/libs/tunnel/src/lib.rs [hunk 1]
```diff
@@ -89,6 +91,7 @@ mod device_channel;
 
 const RESET_PACKET_COUNT_INTERVAL: Duration = Duration::from_secs(1);
 const REFRESH_PEERS_TIMERS_INTERVAL: Duration = Duration::from_secs(1);
+const SENTINEL_DNS_SERVER: Ipv4Addr = Ipv4Addr::new(100, 100, 111, 1);
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/lib.rs]: This is already defined elsewhere as `DNS_SENTINEL`

File: rust/connlib/libs/tunnel/src/lib.rs [hunk 2]
```diff
@@ -224,17 +227,7 @@ where
...
-            let mut resources = self.resources.write();
-            resources.insert(resource_description);
-            resources.resource_list()
-        };
-        match resource_list {
-            Ok(resource_list) => {
-                self.callbacks.on_update_resources(resource_list);
-            }
-            Err(err) => self.callbacks.on_error(&err.into(), Fatal),
-        }
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/lib.rs]: Why was this removed?

File: rust/connlib/libs/tunnel/src/lib.rs [hunk 3]
```diff
@@ -250,13 +243,14 @@ where
...
                 .expect("Couldn't initiate interface");
+            iface_config
+                .add_route(&SENTINEL_DNS_SERVER.into(), self.callbacks())
+                .await?;
         }
 
         self.start_timers();
         self.start_iface_handler();
 
-        self.callbacks.on_tunnel_ready();
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/lib.rs]: Why was this removed?

File: rust/connlib/libs/tunnel/src/ip_packet.rs
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs:1]: There are numerous opportunities to use the `Self` alias throughout this module.

Additionally, you've used `then_some` in multiple places instead of `then`, so the argument will be eagerly evaluated even when unused. This is only desirable if the expression is trivial to evaluate.

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        }
+    }
+}
+
+pub(crate) fn to_dns<'a>(pkt: &'a UdpPacket<'a>) -> Option<&'a Message<[u8]>> {
+    if pkt.get_destination() == DNS_PORT {
+        Message::from_slice(pkt.payload()).ok()
+    } else {
+        None
+    }
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs]:
```suggestion
    (pkt.get_destination() == DNS_PORT)
        .then(|| Message::from_slice(pkt.payload()).ok())
        .flatten()
```

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    pub(crate) fn swap_src_dst(&mut self) {
+        match self {
+            MutableIpPacket::MutableIpv4Packet(p) => {
+                swap_src_dst!(p);
+            }
+            MutableIpPacket::MutableIpv6Packet(p) => {
+                swap_src_dst!(p);
+            }
+        }
+    }
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs:75]:
```suggestion
    }
    
```

**Review by francesca64 — CHANGES_REQUESTED** [review]

File: rust/connlib/libs/tunnel/src/lib.rs (hunk 2 above)
Comment (conectado) [reply rust/connlib/libs/tunnel/src/lib.rs]: My bad, this happened when rebasing and I didn't see it.

File: rust/connlib/libs/tunnel/src/lib.rs (hunk 3 above)
Comment (conectado) [reply rust/connlib/libs/tunnel/src/lib.rs]: same as above

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        $p.set_source(dst);
+        $p.set_destination(src);
+    };
+}
+
+impl<'a> MutableIpPacket<'a> {
+    pub(crate) fn new(data: &mut [u8]) -> Option<MutableIpPacket> {
+        match data[0] >> 4 {
+            4 => MutableIpv4Packet::new(data).map(MutableIpPacket::from),
+            6 => MutableIpv6Packet::new(data).map(MutableIpPacket::from),
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs]:
```suggestion
    pub(crate) fn new(data: &mut [u8]) -> Option<Self> {
        match data[0] >> 4 {
            4 => MutableIpv4Packet::new(data).map(Into::into),
            6 => MutableIpv6Packet::new(data).map(Into::into),
```

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+pub(crate) enum IpPacket<'a> {
+    Ipv4Packet(Ipv4Packet<'a>),
+    Ipv6Packet(Ipv6Packet<'a>),
+}
+
+impl<'a> IpPacket<'a> {
+    pub(crate) fn new(data: &[u8]) -> Option<IpPacket> {
+        match data[0] >> 4 {
+            4 => Ipv4Packet::new(data).map(IpPacket::from),
+            6 => Ipv6Packet::new(data).map(IpPacket::from),
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs]:
```suggestion
    pub(crate) fn new(data: &[u8]) -> Option<Self> {
        match data[0] >> 4 {
            4 => Ipv4Packet::new(data).map(Into::into),
            6 => Ipv6Packet::new(data).map(Into::into),
```

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            _ => None,
+        }
+    }
+
+    pub(crate) fn next_header(&self) -> IpNextHeaderProtocol {
+        match self {
+            IpPacket::Ipv4Packet(p) => p.get_next_level_protocol(),
+            IpPacket::Ipv6Packet(p) => p.get_next_header(),
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs]:
```suggestion
            Self::Ipv4Packet(p) => p.get_next_level_protocol(),
            Self::Ipv6Packet(p) => p.get_next_header(),
```

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    fn from(pkt: Ipv4Packet<'a>) -> Self {
+        Self::Ipv4Packet(pkt)
+    }
+}
+
+impl<'a> From<Ipv6Packet<'a>> for IpPacket<'a> {
+    fn from(pkt: Ipv6Packet<'a>) -> Self {
+        Self::Ipv6Packet(pkt)
+    }
+}
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs:205]:
```suggestion
}

```

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        match self {
+            Self::MutableIpv4Packet(p) => p.to_immutable().into(),
+            Self::MutableIpv6Packet(p) => p.to_immutable().into(),
+        }
+    }
+
+    pub(crate) fn as_udp(&mut self) -> Option<MutableUdpPacket> {
+        self.to_immutable()
+            .is_udp()
+            .then_some(MutableUdpPacket::new(self.payload_mut()))
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs]:
```suggestion
            .then(|| MutableUdpPacket::new(self.payload_mut()))
```

File: rust/connlib/libs/tunnel/src/ip_packet.rs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+            .flatten()
+    }
+
+    pub(crate) fn as_immutable_udp(&self) -> Option<UdpPacket> {
+        self.to_immutable()
+            .is_udp()
+            .then_some(UdpPacket::new(self.payload()))
```
Comment (francesca64) [inline rust/connlib/libs/tunnel/src/ip_packet.rs]:
```suggestion
            .then(|| UdpPacket::new(self.payload()))
```

**Review by francesca64 — CHANGES_REQUESTED** [review]

File: rust/connlib/libs/tunnel/src/ip_packet.rs (hunk 6 above)
Comment (conectado) [reply rust/connlib/libs/tunnel/src/ip_packet.rs]: I rather not use a `Self` return type here, that disables the elision and requires a `&'a` for the `data parameter.

Comment (francesca64) [reply rust/connlib/libs/tunnel/src/ip_packet.rs]: Ah, I hadn't considered that, fair enough!

**Review by francesca64 — APPROVED** [review]

**Review by AndrewDryga — APPROVED** [review]
